### PR TITLE
JDK-8304959: Public API in javafx.css.Match should not return private API class PseudoClassState

### PR DESCRIPTION
**Pull request by hjohn**
The class `PseudoClassState` is private API, but was exposed erroneously in the CSS API. Instead, `Set<PseudoClass>` should have been used. This PR corrects this.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change requires CSR request [JDK-8304994](https://bugs.openjdk.org/browse/JDK-8304994) to be approved
- [x] Change must be properly reviewed (2 reviews required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer), 1 [Author](https://openjdk.org/bylaws#author))

### Issues
 * [JDK-8304959](https://bugs.openjdk.org/browse/JDK-8304959): Public API in javafx.css.Match should not return private API class PseudoClassState
 * [JDK-8304994](https://bugs.openjdk.org/browse/JDK-8304994): Public API in javafx.css.Match should not return private API class PseudoClassState (**CSR**)

### Reviewers
 * [Kevin Rushforth](https://openjdk.org/census#kcr) (@kevinrushforth - **Reviewer**)
 * [Andy Goryachev](https://openjdk.org/census#angorya) (@andy-goryachev-oracle - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jfx.git pull/1070/head:pull/1070` \
`$ git checkout pull/1070`

Update a local copy of the PR: \
`$ git checkout pull/1070` \
`$ git pull https://git.openjdk.org/jfx.git pull/1070/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 1070`

View PR using the GUI difftool: \
`$ git pr show -t 1070`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jfx/pull/1070.diff">https://git.openjdk.org/jfx/pull/1070.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jfx/pull/1070#issuecomment-1485219910)

### CODE REVIEW

**Comment by kevinrushforth** [comment]
This was clearly a mistake back when the public API was first created in JDK 9.

/reviewers 2
/csr

File: modules/javafx.graphics/src/main/java/javafx/css/Match.java [hunk 1]
```diff
@@ -77,7 +76,7 @@ public Selector getSelector() {
      * Gets the pseudo class state.
      * @return the pseudo class state
      */
-    public PseudoClassState getPseudoClasses() {
+    public Set<PseudoClass> getPseudoClasses() {
```
Comment (andy-goryachev-oracle) [inline modules/javafx.graphics/src/main/java/javafx/css/Match.java:79]: does this change need a CSR?

Comment (kevinrushforth) [reply modules/javafx.graphics/src/main/java/javafx/css/Match.java:79]: Yes, thanks for checking. I already flagged it as needing a CSR.

**Comment by hjohn** [comment]
@kevinrushforth I've created the CSR for this.  Can I move it to proposed?

Also, I've been looking into https://bugs.openjdk.org/browse/JDK-8199216 which sees excessive use of `PseudoClassState` instances (and I see this even under more "normal" circumstances as well).  I've got a fix for this issue which involves making use of an immutable variant of `PseudoClassState` (as almost 99% of all the instances are never mutated, but must be copied just in case they are). 

This change would need to go in first though to make this possible.

**Comment by kevinrushforth** [comment]
> @kevinrushforth I've created the CSR for this. Can I move it to proposed?

I took a quick look, and it is fine to move to Proposed.

File: modules/javafx.graphics/src/main/java/javafx/css/Match.java [hunk 2]
```diff
@@ -41,15 +40,15 @@
 public final class Match implements Comparable<Match> {
 
     final Selector selector;
-    final PseudoClassState pseudoClasses;
+    final Set<PseudoClass> pseudoClasses;
```
Comment (kevinrushforth) [inline modules/javafx.graphics/src/main/java/javafx/css/Match.java:43]: Should this be an `ObservableSet`? Changing the type to the `Set` superclass will mean that applications would need to do an `instanceof` check to know whether it was observable or not?

**Review by kevinrushforth — COMMENTED** [review]
I did leave one question that you might want to address before proposing the CSR.

File: modules/javafx.graphics/src/main/java/javafx/css/Match.java (hunk 2 above)
Comment (kevinrushforth) [reply modules/javafx.graphics/src/main/java/javafx/css/Match.java:43]: Btw, I meant to add this comment on the public getter method.

File: modules/javafx.graphics/src/main/java/javafx/css/Match.java (hunk 1 above)
Comment (kevinrushforth) [inline modules/javafx.graphics/src/main/java/javafx/css/Match.java:79]: Should this be an `ObservableSet`? Changing the type to the `Set` superclass will mean that applications would need to do an `instanceof` check to know whether it was observable or not?

Comment (hjohn) [reply modules/javafx.graphics/src/main/java/javafx/css/Match.java:79]: I think that `Match` is supposed to be immutable, given the non-public constructor.  Match itself will never change the set (and nothing else will) so making it observable seems unnecessary.

However, this was not correctly spec'd, and you can change a `Match` now, and since it doesn't make a copy, you'd be changing the pseudo classes of whatever selector created it.  Note that `SimpleSelector` does make a copy, and goes to great pains to not expose anything mutable, but exposes it accidentally via `Match`.

In other words, `simpleSelector.createMatch().getPseudoClasses().clear()` would break the Selectors encapsulation.

I think it's best to close that loophole.  If you agree, I can document this method that it returns an immutable set, which is also what I assumed would be the case in my other PR where I made many of these immutable.

Comment (hjohn) [reply modules/javafx.graphics/src/main/java/javafx/css/Match.java:79]: The CSS API baffles me a bit.  It doesn't seem consistent.

Just now I took a look at the class `SimpleSelector` and `CompoundSelector`. These are public, yet cannot be constructed by users.  They're also not returned anywhere (the closest is `Selector#createSelector` which returns the interface).

Essentially this means that `SimpleSelector` and `CompoundSelector` should probably be package private.  Yet, I guess they were made public because `SelectorPartitioning` is doing instanceof checks and is casting to these classes.  But anybody can do that now, and that means that for example `SimpleSelector#getStyleClassSet` is exposed, which returns a mutable set...

Reading between the lines though it seems that `SimpleSelector` and `CompoundSelector` were intended to be fully immutable (which makes sense as they represent a style sheet).  Any changes would not be picked up as nothing is observing these properties.

I think these loopholes should be closed.

There are two options IMHO:

1. Move `SimpleSelector` and `CompoundSelector` to the com hierarchy.  They can't be publically constructed, and are never returned.  The only way to reach them is by casting.
2. If it's too late for that, then close all loopholes and ensure that these two classes are fully immutable. From what I can see now, only `getStyleClassSet` and the mentioned method in `Match` need closing.  `CompoundSelector` is already immutable.

Comment (kevinrushforth) [reply modules/javafx.graphics/src/main/java/javafx/css/Match.java:79]: I'll take a closer look early next week.

Comment (kevinrushforth) [reply modules/javafx.graphics/src/main/java/javafx/css/Match.java:79]: > I think that Match is supposed to be immutable, given the non-public constructor. Match itself will never change the set (and nothing else will) so making it observable seems unnecessary.

Agreed.

> In other words, simpleSelector.createMatch().getPseudoClasses().clear() would break the Selectors encapsulation.
> 
> I think it's best to close that loophole. If you agree, I can document this method that it returns an immutable set, which is also what I assumed would be the case in my other PR where I made many of these immutable.

Yes, this seems like the best solution to me.

> Essentially this means that SimpleSelector and CompoundSelector should probably be package private. 

I agree that these two classes should not have been made public when the javafx.css package was created as public API back in JDK 9.

The usual process for removing API is to deprecate it for removal in one release and then remove it in a future release, but in this case, since those classes cannot be constructed, and are never returned by any public API, any use of them would be relying on an implementation detail (via an `instanceof` and a cast, to no good purpose).

Since there is no useful way an application could be using these classes, I recommend option 1, as long as those two classes can be cleanly moved to com.sun.javafx.css. Otherwise, we could go with option 2 along with deprecating those two classes for removal.

The Specification section of the CSR would simply propose to remove those two classes. You could describe what is happening (moving them to a non-exported package) in the Solution section.

Comment (hjohn) [reply modules/javafx.graphics/src/main/java/javafx/css/Match.java:79]: @kevinrushforth It looks like it will have to be option 2; the reason is that both `Selector` and `Match` are abstract classes (with a package private constructor).  The `CompoundSelector` and `SimpleSelector` extend this abstract class (as there is some small overlap). If these had been interfaces, it would have allowed this.  

I couldn't find anything about binary compatibility when changing an abstract class with a package private constructor to an interface, so I tested it and I got an `IncompatibleClassChangeError` when calling `Selector Selector.createSelector(...)` method.  Apparently, the compiled method call does encode the difference between an interface and a class:

    invokestatic  #7                  // Method Selector.createSelector:(I)LSelector;

vs:

    invokestatic  #7                  // InterfaceMethod Selector.createSelector:(I)LSelector;

At the most, I could make `CompoundSelector` and `SimpleSelector` nested classes in `Selector`; they would at least be hidden then.

**Comment by hjohn** [comment]
I think this is the best that can be done for now.  There are still some mutable aspects to `Selector` (you can change the `Rule` and the ordinal of one).  I'm really not quite sure why it is done that way, it seems there was may have been some vision that these classes could be used to "edit" a stylesheet -- something that's probably best done with a different set of classes to avoid burdening regular read-only CSS work with editable variants, requiring defensive copying everywhere.

Ideally, `Selector` and `Match` would have been interfaces, and the private code should work with the CSS classes only through those.  Changing them to interfaces now seems impossible, it's not binary compatible, and I don't see a deprecation route to get there.

However, it is still possible to improve the situation a bit (in another PR perhaps), as the private code only requires a few additional methods to be made public API on `Selector` to remove its direct dependencies on `SimpleSelector` and `CompoundSelector`:

- getName() which returns the java class name to which the selector is applied
- getId() returns the CSS id (if any) of a Selector
- getStyleClassSet() returns the (immutable) style classes of a Selector

These three methods are implemented by `SimpleSelector`, but not by `CompoundSelector`.  Instead, its last selector is extracted (which is always a `SimpleSelector`) and the methods are called on those. These methods could be made part of `Selector`, then `CompoundSelector` implements them by directing them to its last `SimpleSelector`, and the private code would not need to care about the selector type anymore. These selector classes could then be hidden completely.

**Comment by hjohn** [comment]
I had to pull in some of the fixes I did in `BitSet` in https://github.com/openjdk/jfx/pull/1076

Just wrapping a `Set<PseudoClass>` in an unmodifiable set is enough to break lots of tests, because `BitSet` has many methods that will simply do nothing if the "other" `Set` is not a `BitSet`.

**Comment by hjohn** [comment]
@kevinrushforth about the CSR -- as I couldn't move the classes to non-public hierarchy, I think the CSR is currently correct.  I also removed a public method in `SimpleSelector` as it wasn't used (and as discussed before, nobody can reach it without casting).

There were a few javadoc only changes that could be seen as clarification (whether something can be `null` or whether it is immutable).  I could include those in the CSR.

If we want to move forward later to perhaps try and remove `SimpleSelector` and `CompoundSelector` (perhaps by adding those few public API methods to `Selector`), I could already mark these deprecated (or I could do so to indicate they shouldn't be used at all, whether we want to remove them or not).  What do you think?

**Comment by hjohn** [comment]
@kevinrushforth do you think we can move forward with this?

**Comment by andy-goryachev-oracle** [comment]
I'll take a look at this next week.

File: modules/javafx.graphics/src/main/java/javafx/css/SimpleSelector.java [hunk 3]
```diff
@@ -70,23 +70,7 @@ public String getName() {
     }
 
     /**
-     * Gets an immutable list of style-classes of the {@code Selector}.
-     * @return an immutable list of style-classes of the {@code Selector}
-     */
-    public List<String> getStyleClasses() {
```
Comment (kevinrushforth) [inline modules/javafx.graphics/src/main/java/javafx/css/SimpleSelector.java:76]: This is a public method, so removing it would be an incompatible API change. Given the decision to live with this public class, I think you should restore this method.

File: modules/javafx.graphics/src/main/java/javafx/css/SimpleSelector.java [hunk 4]
```diff
@@ -168,41 +130,42 @@ public NodeOrientation getNodeOrientation() {
         // then match needs to check name
         this.matchOnName = (name != null && !("".equals(name)) && !("*".equals(name)));
 
-        this.styleClassSet = new StyleClassSet();
+        Set<StyleClass> scs = new StyleClassSet();
 
-        int nMax = styleClasses != null ? styleClasses.size() : 0;
-        for(int n=0; n<nMax; n++) {
+        if (styleClasses != null) {
+            for(int n = 0; n < styleClasses.size(); n++) {
```
Comment (kevinrushforth) [inline modules/javafx.graphics/src/main/java/javafx/css/SimpleSelector.java:136]: Minor: space after `for`

File: modules/javafx.graphics/src/main/java/javafx/css/SimpleSelector.java [hunk 5]
```diff
@@ -168,41 +130,42 @@ public NodeOrientation getNodeOrientation() {
...
+        this.styleClassSet = Collections.unmodifiableSet(scs);
         this.matchOnStyleClass = (this.styleClassSet.size() > 0);
 
-        this.pseudoClassState = new PseudoClassState();
+        PseudoClassState pcs = new PseudoClassState();
+        NodeOrientation dir = NodeOrientation.INHERIT;
 
-        nMax = pseudoClasses != null ? pseudoClasses.size() : 0;
+        if (pseudoClasses != null) {
+            for(int n = 0; n < pseudoClasses.size(); n++) {
```
Comment (kevinrushforth) [inline modules/javafx.graphics/src/main/java/javafx/css/SimpleSelector.java:152]: Minor: space after `for`

**Review by kevinrushforth — COMMENTED** [review]
The changes look good with a few inline comments.

File: modules/javafx.graphics/src/main/java/com/sun/javafx/css/BitSet.java [hunk 6]
```diff
@@ -493,33 +515,24 @@ public void clear() {
...
-            for (int n = 0; n < bits.length; n++) {
-                final long mask = bits[n];
-                hash = 71 * hash + (int)(mask ^ (mask >>> 32));
-            }
-        }
-        return hash;
-    }
-
     @Override
     public boolean equals(Object obj) {
```
[the remark this answers is not in the document]
Comment (hjohn) [reply modules/javafx.graphics/src/main/java/com/sun/javafx/css/BitSet.java:526]: I made it explicit, and slightly reworded the comment. Please note, the equals code that BitSet uses is still broken in this PR, but as fixing it entails more changes to BitSet, and the goal of this PR is to fix a different issue, I've left that for #1076

File: modules/javafx.graphics/src/main/java/javafx/css/Match.java [hunk 7]
```diff
@@ -67,17 +70,19 @@ public final class Match implements Comparable<Match> {
 
     /**
      * Gets the {@code Selector}.
-     * @return the {@code Selector}
+     *
+     * @return the {@code Selector}, never {@code null}
```
[the remark this answers is not in the document]
Comment (hjohn) [reply modules/javafx.graphics/src/main/java/javafx/css/Match.java:74]: Added both changes to the CSR.

File: modules/javafx.graphics/src/main/java/com/sun/javafx/css/BitSet.java [hunk 8]
```diff
@@ -229,10 +230,14 @@ public boolean contains(Object o) {
     /** {@inheritDoc} */
     @Override
     public boolean containsAll(Collection<?> c) {
+        if (this.getClass() != c.getClass()) {
```
Comment (andy-goryachev-oracle) [inline modules/javafx.graphics/src/main/java/com/sun/javafx/css/BitSet.java:233]: this change is not equivalent: in the old code, null `c` would return false, in the new it'll throw an NPE

File: modules/javafx.graphics/src/main/java/com/sun/javafx/css/BitSet.java [hunk 9]
```diff
@@ -258,10 +263,14 @@ public boolean containsAll(Collection<?> c) {
     /** {@inheritDoc} */
     @Override
     public boolean addAll(Collection<? extends T> c) {
+        if (this.getClass() != c.getClass()) {
```
Comment (andy-goryachev-oracle) [inline modules/javafx.graphics/src/main/java/com/sun/javafx/css/BitSet.java:266]: same, handling of the null 'c' argument.
I wonder if this is intended behavior?

File: modules/javafx.graphics/src/main/java/com/sun/javafx/css/BitSet.java [hunk 10]
```diff
@@ -330,10 +339,19 @@ public boolean addAll(Collection<? extends T> c) {
     /** {@inheritDoc} */
     @Override
     public boolean retainAll(Collection<?> c) {
+        if (this.getClass() != c.getClass()) {
```
Comment (andy-goryachev-oracle) [inline modules/javafx.graphics/src/main/java/com/sun/javafx/css/BitSet.java:342]: ditto, null 'c'

File: modules/javafx.graphics/src/main/java/com/sun/javafx/css/BitSet.java [hunk 11]
```diff
@@ -412,10 +430,14 @@ public boolean retainAll(Collection<?> c) {
     /** {@inheritDoc} */
     @Override
     public boolean removeAll(Collection<?> c) {
+        if (this.getClass() != c.getClass()) {
```
Comment (andy-goryachev-oracle) [inline modules/javafx.graphics/src/main/java/com/sun/javafx/css/BitSet.java:433]: null 'c'

File: modules/javafx.graphics/src/main/java/com/sun/javafx/css/BitSet.java [hunk 12]
```diff
@@ -495,31 +517,28 @@ public void clear() {
...
-        if (this == obj) {
+        // Note: overridden to provide a fast path; both equals and hashCode MUST respect
+        // the Set contract to interact correctly with sets of other types!
+        if (obj == this) {
             return true;
         }
-
-        if (obj == null || this.getClass() != obj.getClass()) {
-            return false;
+        if (obj instanceof BitSet<?> bitSet) {  // fast path if other is a BitSet
```
Comment (andy-goryachev-oracle) [inline modules/javafx.graphics/src/main/java/com/sun/javafx/css/BitSet.java:532]: I think we may need to add a 'component type' to BitSet, to make sure than BitSet<A> never intersects with BitSet<B>.

(Didn't we do it already, may be in some other PR?)

File: modules/javafx.graphics/src/main/java/javafx/css/Match.java [hunk 13]
```diff
@@ -40,24 +41,26 @@
...
     final int styleClassCount;
+    final int idCount;
 
     // CSS3 spec gives weight to id count, then style class count,
     // then pseudoclass count, and finally matching types (i.e., java name count)
     final int specificity;
 
-    Match(final Selector selector, PseudoClassState pseudoClasses, int idCount, int styleClassCount) {
-        assert selector != null;
+    Match(final Selector selector, Set<PseudoClass> pseudoClasses, int idCount, int styleClassCount) {
```
Comment (andy-goryachev-oracle) [inline modules/javafx.graphics/src/main/java/javafx/css/Match.java:54]: just curious: do we really need a 'final' for an effectively final argument?

File: modules/javafx.graphics/src/main/java/javafx/scene/CssStyleHelper.java [hunk 14]
```diff
@@ -131,7 +131,10 @@ static CssStyleHelper createStyleHelper(final Node node) {
                 node.styleHelper.cacheContainer.fontSizeCache.clear();
             }
             node.styleHelper.cacheContainer.forceSlowpath = true;
-            node.styleHelper.triggerStates.addAll(triggerStates[0]);
+
+            if (triggerStates[0] != null) {
```
Comment (andy-goryachev-oracle) [inline modules/javafx.graphics/src/main/java/javafx/scene/CssStyleHelper.java:135]: I see that a null check was added.
Are we sure we added it to *every* possible place?

(it's not trivial to verify that with Eclipse without massive changes)

**Review by andy-goryachev-oracle — CHANGES_REQUESTED** [review]
two sticky points:
- verify somehow that nulls cannot be passed to *All() methods
- consider adding a component type to BitSet

File: modules/javafx.graphics/src/main/java/com/sun/javafx/css/BitSet.java (hunk 8 above)
Comment (hjohn) [reply modules/javafx.graphics/src/main/java/com/sun/javafx/css/BitSet.java:233]: That's correct, before it broke the contract of `Set`.  It's not supposed to return `false` when the parameter is `null`.  This class is sometimes exposed as public API (via `ObservableSet`) and so must respect the `Set` contract.

File: modules/javafx.graphics/src/main/java/com/sun/javafx/css/BitSet.java (hunk 9 above)
Comment (hjohn) [reply modules/javafx.graphics/src/main/java/com/sun/javafx/css/BitSet.java:266]: I doubt the `Set` contract was breached here on purpose. More likely, the original implementation was never exposed, or the original developer didn't realize that `Set` has a very specific contract that you must follow, or it won't interact well with other collections.

File: modules/javafx.graphics/src/main/java/com/sun/javafx/css/BitSet.java (hunk 12 above)
Comment (hjohn) [reply modules/javafx.graphics/src/main/java/com/sun/javafx/css/BitSet.java:532]: We did, in #1076. However, I would like to not pull in all the `BitSet` fixes into this PR as this has nothing to do with `BitSet` and its many problems.

Comment (andy-goryachev-oracle) [reply modules/javafx.graphics/src/main/java/com/sun/javafx/css/BitSet.java:532]: Since it's an internal class, and both PseudoClassState and StyleClassSet are final, checking for class == class is probably sufficient.
Do you think we should add a warning to BitSet, it is not a problem since no-one will ever extend it?

File: modules/javafx.graphics/src/main/java/javafx/css/Match.java (hunk 13 above)
Comment (hjohn) [reply modules/javafx.graphics/src/main/java/javafx/css/Match.java:54]: `final` here will just warn you if you try to reassign the parameter.  It was present in the original code.

Reassigning locals and parameters should generally be avoided. For the first, you can have your IDE generate a warning or error.  For the second, I have the habit of always assigning a variable when it is declared.  In the rare cases where this can't be done, the `final` keyword offers little benefit.

The keyword does not offer any further benefits on parameters and locals, and it hasn't since Java added the concept of effectively final (in Java 8 I think).  Before this, you sometimes needed to add `final` so a local could be referred to in anonymous subclasses.

File: modules/javafx.graphics/src/main/java/javafx/scene/CssStyleHelper.java (hunk 14 above)
Comment (hjohn) [reply modules/javafx.graphics/src/main/java/javafx/scene/CssStyleHelper.java:135]: The underlying collection `BitSet` is only used by classes dealing with styles and pseudo class states. I checked the classes involved, and only this one needed changing.  The check was added because `BitSet` erroneously allowed adding `null` before (which resulted in no change), but that breaks the `Set` contract.

In any case, adding `null` is an error, we should never be lenient when it comes to `null`s.

File: modules/javafx.graphics/src/main/java/com/sun/javafx/css/BitSet.java (hunk 9 above)
Comment (hjohn) [reply modules/javafx.graphics/src/main/java/com/sun/javafx/css/BitSet.java:266]: To be clear, `BitSet` is currently exposed via `Node#getPseudoClassStates`.

I suspect it is almost never used by anyone, because there are a lot of bugs.

For example, equality won't work correctly:

     node.getPseudoClassStates().equals(Set.of( ... ));  // Always false, sets are of different types

Nor will `containsAll`:

     node.getPseudoClassStates().containsAll(Set.of( ... ));  // Always false, because Set is not a BitSet
     node.getPseudoClassStates().containsAll(null);  // Doesn't throw NPE, returns false

All of these are `Set` contract violations, that I suppose were thought to be irrelevant because it is a com.sun.* class.

File: modules/javafx.graphics/src/main/java/com/sun/javafx/css/BitSet.java (hunk 12 above)
Comment (hjohn) [reply modules/javafx.graphics/src/main/java/com/sun/javafx/css/BitSet.java:532]: This is also solved in #1076:

     if (obj instanceof BitSet<?> bitSet && getElementType().equals(bitSet.getElementType()))

But since that requires pulling in even more changes (`getElementType`) I've left it there.

Relying on subclasses being final though seems like a bad idea for an abstract class, unless we seal it and use permits (or, just remove `BitSet` completely... it's purpose will be minimal if #1076 is accepted).

Comment (andy-goryachev-oracle) [reply modules/javafx.graphics/src/main/java/com/sun/javafx/css/BitSet.java:532]: that's my point.  I suppose we could use sealed class here, since javafx requires minimum java17?

Comment (hjohn) [reply modules/javafx.graphics/src/main/java/com/sun/javafx/css/BitSet.java:532]: Do you think this needs to be changed as part of this PR?

Note that the `==` solution requires:
- sealing BitSet permitting only PseudoClassState and StyleClassSet
- adding a warning that if you ever unseal it, that you must re-evaluate the `equals` implementation
- adding a warning to PseudoClassState / StyleClassSet to re-evaluate `BitSet#equals` if you remove `final`

The `instanceof` solution allows arbitary subclasses, like any other Set implementation does (`AbstractSet#equals` for example use `instanceof Set`) and requires no further changes.

What I'm missing is the reasoning why you would want to change this. I realize that the other solution can also work, but that's not enough IMHO. For me the two options are practically equivalent, neither being truly better than the other, so then it is a matter of preference, and I slightly favor using `instanceof` here to avoid potential problems down the road.

Comment (kevinrushforth) [reply modules/javafx.graphics/src/main/java/com/sun/javafx/css/BitSet.java:532]: > Do you think this needs to be changed as part of this PR?

I'll answer this part of your question: No.

It seems out of scope for this particular bug fix, which is aimed at fixing an API and impl problem with `Match`, and should limit the changes to other classes to those needed to accomplish this.

It's great to point out existing weaknesses in, say, `BitSet`, but they should be addressed separately, after evaluating the need.

**Comment by andy-goryachev-oracle** [comment]
Consider this example:

```
    @Test
    public void testEquals() {
        PseudoClassState a = new PseudoClassState();
        StyleClassSet b = new StyleClassSet();
        assertFalse(a.equals(b));
    }
```

**Comment by hjohn** [comment]
> Consider this example:
> 
> ```
>     @Test
>     public void testEquals() {
>         PseudoClassState a = new PseudoClassState();
>         StyleClassSet b = new StyleClassSet();
>         assertFalse(a.equals(b));
>     }
> ```

I'm aware that this is broken now, and still is broken after this PR.  It's fixed in #1076 and has this test case even.

It's worse than that tough, even `BitSet`s of the same type can fail equals (from #1076):

    @Test
    void shouldBeEqual() {
        StyleClassSet set1 = new StyleClassSet();
        StyleClassSet set2 = new StyleClassSet();

        set1.add(StyleClassSet.getStyleClass("abc"));
        set2.add(StyleClassSet.getStyleClass("abc"));

        assertEquals(set1, set2);

        for (int i = 0; i < 1000; i++) {
            // grow internal bit set array:
            set1.add(StyleClassSet.getStyleClass("" + i));

            assertNotEquals(set1, set2);
        }

        for (int i = 0; i < 1000; i++) {
            set1.remove(StyleClassSet.getStyleClass("" + i));
        }

        // still equal despite internal array sizes being different size:
        assertEquals(set1, set2);
    }

Do you want me to pull in more changes from #1076 unrelated to fixing the public API problem in Match?

**Comment by andy-goryachev-oracle** [comment]
My concern is that replacing == with .equals() in BitSet.equals() may lead to issues in internal CSS code.  It is highly unlikely that application would extend BitSet, or put StyleClassSet in the same hash table due to both being in com.sun; but the internal code might (and I don't want to comb the CSS code looking for possible occurrences).

If we replace line 532 with `if(getClass == obj.getClass()) {` however, we would not have to.

**Comment by kevinrushforth** [comment]
> It is highly unlikely that application would extend BitSet, or put StyleClassSet in the same hash table due to both being in com.sun;

Not merely unlikely, it is not possible, so this need not be considered.

> but the internal code might (and I don't want to comb the CSS code looking for possible occurrences).
>
> If we replace line 532 with `if(getClass == obj.getClass()) {` however, we would not have to.

That's a good suggestion. Given that we want the minimal changes to BitSet for this particular PR, this seems a reasonable, and safe choice. `BitSet` is an internal-only abstract class, and the only two concrete subclasses are both final with different type parameters.

**Comment by hjohn** [comment]
> My concern is that replacing == with .equals() in BitSet.equals() may lead to issues in internal CSS code. It is highly unlikely that application would extend BitSet, or put StyleClassSet in the same hash table due to both being in com.sun; but the internal code might (and I don't want to comb the CSS code looking for possible occurrences).
> 
> If we replace line 532 with `if(getClass == obj.getClass()) {` however, we would not have to.

Alright, I see your point.  The current version will always return `false` when comparing `StyleClassSet` with a `PseudoClassState`.  So does the (full) solution in #1076 -- but this half copied solution doesn't, which is a slight regression (probably harmless, but even so).

I'll fix it.

**Comment by hjohn** [comment]
I will still fix it so it isn't worse than it was before, but I just found another reason why you shouldn't use `getClass() == other.getClass()` when dealing with sets:

    @Test
    void twoEmptyBitSetsShouldBeEqual() {

        /*
         * Per Set contract, the empty set is equal to any other empty set.
         */

        assertEquals(new StyleClassSet(), new PseudoClassState());
        assertEquals(new PseudoClassState(), new StyleClassSet());
        assertEquals(Set.of(), new PseudoClassState());
        assertEquals(new PseudoClassState(), Set.of());
        assertEquals(Set.of(), new StyleClassSet());
        assertEquals(new StyleClassSet(), Set.of());
    }

**Comment by hjohn** [comment]
Sorry for the force push, that wasn't what I intended; wanted to undo changes GitHub did on the remote by merging in master as I also merged it in locally (but I added a commit just before).  Unfortunately, the PR already picked up the master merge from GitHub... I'll know that for next time.

**Review by andy-goryachev-oracle — APPROVED** [review]
tested with the MonkeyTester on macOS Ventura 13.3.1(a), including dynamic re-loading of the stylesheets.  see no ill effects.

**Comment by kevinrushforth** [comment]
I took a closer look at the compatibility issues Joe raised in the CSR, and I can confirm that yes, this change will break binary compatibility for any application that calls the existing method. It would also can break source compatibility for the subset of those applications that use the return value as an `ObservableSet` (which is the lowest public supertype of `PseudoClassState`), although this latter concern seems _very_ unlikely, and not all that impactful.

So it's the binary compatibility we need to consider. The question is: is the `Match::getPseudoClasses` method being called in a useful manner by applications today?

If not, then the change seems OK.

If it is, then we might need to consider adding a replacement method and deprecating the existing method for removal. The implementation of the replacement method would be exactly what your PR currently has for the existing method, and the existing method, would create a `PseudoClassState` object to hold a copy of the objects in the set.

**Comment by hjohn** [comment]
> So it's the binary compatibility we need to consider. The question is: is the `Match::getPseudoClasses` method being called in a useful manner by applications today?
> 
> If not, then the change seems OK.
> 
> If it is, then we might need to consider adding a replacement method and deprecating the existing method for removal. The implementation of the replacement method would be exactly what your PR currently has for the existing method, and the existing method, would create a `PseudoClassState` object to hold a copy of the objects in the set.

I've done some Googling with `"getPseudoClasses" import javafx`, and also looked at several code search sites, and if it's used, then I can't find any examples at all.  I also searched for `createMatch`, with similar results, no uses outside the JavaFX code base.

To call this method, one would to do something like this at a minimum:

     StyleSheet styleSheet = StyleSheet.loadBinary( <input stream> );

     for (Rule rule : styleSheet.getRules()) {
          for (Selector selector : rule.getSelectors()) {
               Match match = selector.createMatch();
 
               match.getPseudoClassStates();  // here...
          }
     }

I did find a few references to `loadBinary`, but nothing for `getRules` which would be needed next.

**Comment by kevinrushforth** [comment]
> I've done some Googling with `"getPseudoClasses" import javafx`, and also looked at several code search sites, and if it's used, then I can't find any examples at all. I also searched for `createMatch`, with similar results, no uses outside the JavaFX code base.
> ...
> There's not a lot of reason to call `getPseudoClassStates` -- the method is mainly public so internal code outside its the public package can reach these to do the actual matching logic.

This is what I would have expected as well. I'll do some additional searching as well, but, as of now, I think we can proceed down the current path and live with the incompatibility.

**Review by kevinrushforth — APPROVED** [review]
I've completed my review, and everything looks good. I found one missing doc change in the CSR (the change in SimpleSelector) that should be added, and then I'll formally review the CSR.

**Comment by kevinrushforth** [comment]
@hjohn In case you missed seeing it, the CSR was approved, so Skara marked this PR as `ready`.
